### PR TITLE
Add platform check for ownership of objects

### PR DESCRIPTION
**Pull request by def-**
Part of https://github.com/MaterializeInc/materialize/issues/17983

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: misc/python/materialize/checks/owners.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+from materialize.checks.checks import Check
+from materialize.util import MzVersion
+
+
+class Owners(Check):
+    def _can_run(self) -> bool:
+        return self.base_version >= MzVersion.parse("0.47.0-dev")
+
+    def initialize(self) -> Testdrive:
+        return Testdrive("")
```
Comment (philip-stoev) [inline misc/python/materialize/checks/owners.py:22]: The general philosophy of the platform checks framework is that you create a dependency tree between objects and then make sure you create/populate the objects from the distinct stages of the dependency graph at each of the phases. This ensures that objects created in a prior phase (that is, in a prior Mz version or prior to some Mz restart, or in the future backup/restore) can still be used to create further objects in subsequent phases.

In this particular case, the dependency graph seems to be like this:
1. CREATE ROLE
2. login as role + CREATE some object
3. CREATE another object that references first object, perform DDL on first object, perform DML on first object, etc.

If none of the items in #3A are currently on the table, then the dependency graph is much simpler , just items 1 and 2 . Those two items then should be happening in different phases of the test.

Therefore, to overly-exercise it, I would suggest something like this:

* initialize:
- create role r1
* manipulate phase 1
- login as r1 , create object as **r1**
- create role r2
* manipulate phase 2
- login as r1, create object as r1
- login as r2, create object as **r2**
- create role r3, login as r3, create object as r3
* validate (can run several times, so must be idempotent)
- validate ownership of each of the previously-created objects
- login as each prior role, create object, check its ownership, drop it

I think you have most of that stuff already, except reusing roles from prior phases in subsequent ones.

There is quite a bit of a potential for a combinatorial explosion here, especially if item No 3 is also going to land in the future, so there is no expecation for complete exhaustiveness.

File: misc/python/materialize/checks/owners.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        return [
+            Testdrive(dedent(s))
+            for s in [
+                """
+                > CREATE ROLE owner_role_01;
+                $ postgres-execute connection=postgres://owner_role_01@materialized:6875/materialize
+                CREATE TABLE owner_table1 (c1 int, c2 int)
+                CREATE INDEX owner_index1 ON owner_table1 (c2)
+                CREATE VIEW owner_v1 AS SELECT * FROM owner_table1
+                CREATE MATERIALIZED VIEW owner_mv1 AS SELECT * FROM owner_table1
```
Comment (philip-stoev) [inline misc/python/materialize/checks/owners.py:34]: `CREATE SOURCE`, even if `GENERATOR` should also be created at some point -- I can bet 100 euro that there are approximately gazillon paths in the code where `CREATE TABLE` and `CREATE SOURCE` are handled differently.

`CREATE SINK` too.

As `CREATE SOURCE` , `CREATE SINK` is expensive, so creating one of each the entire test would be sufficient.

File: misc/python/materialize/checks/owners.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                """,
+                """
+                > CREATE ROLE owner_role_02;
+                $ postgres-execute connection=postgres://owner_role_02@materialized:6875/materialize
+                CREATE TABLE owner_table2 (c1 int, c2 int)
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply misc/python/materialize/checks/owners.py:39]: Yes, the full list is:

- sinks
- indexes
- connections
- types
- functions
- secrets
- tables
- sources
- views
- materialized views
- databases
- schemas
- clusters
- cluster replicas

[the remark this answers is not in the document]
Comment (jkosh44) [reply misc/python/materialize/checks/owners.py:39]: > I do not think functions can currently be created.

That's correct.

**Comment by def-** [comment]
The test got a lot larger, but the additional objects found 3 missing owners, so good in total: https://github.com/MaterializeInc/materialize/issues/18414

**Comment by def-** [comment]
Cancelled builds because of urgent deploy, but they were looking fine in nightly, locally too.

**Review by philip-stoev — APPROVED** [review]
Thank you!

**Pull request merged**